component/libfuse/native_file_io.h

```diff
@@ -0,0 +1,229 @@
+    if (fobj) {
+        memset(fobj, 0, sizeof(file_handle_t));
+        fobj->fd = fd;
+        fobj->obj = obj;
+    }
+
+    return fobj;
+}
+
+// release_native_file_object : Release the native C-struct for handle 
+static void release_native_file_object(fuse_file_info_t* fi)
+{
+    // Called on close operation from libfuse component
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    if (handle_obj) {
+        free(handle_obj);
+    }
+}
+
+
+// native_pread :  Do pread on file directly without involving any Go code
+static int native_pread(char *path, char *buf, size_t size, off_t offset, file_handle_t* handle_obj)
+{
+    errno = 0;
+    int res = pread(handle_obj->fd, buf, size, offset);
+    if (res == -1)
+        res = -errno;
+    
+    #if 0
+    handle_obj->cnt++;
+    if (!(handle_obj->cnt % CACHE_UPDATE_COUNTER)) {
+        // Time to send a call up to update the cache
+        blobfuse_cache_update(path);
+        handle_obj->cnt = 0;
+    }   
+    #endif
+    
+    return res;
+}
+
+// native_pwrite :  Do pwrite on file directly without involving any Go code
+static int native_pwrite(char *path, char *buf, size_t size, off_t offset, file_handle_t* handle_obj)
+{
+    errno = 0;
+    int res = pwrite(handle_obj->fd, buf, size, offset);
+    if (res == -1)
+        res = -errno;
+
+    // Increment the operation counter and mark a write was done on this handle
+    handle_obj->dirty = 1;
+    handle_obj->cnt++;
+    if (!(handle_obj->cnt % CACHE_UPDATE_COUNTER)) {
+        // Time to send a call up to update the cache
+        blobfuse_cache_update(path);
+        handle_obj->cnt = 0;
+    }
+
+    return res;
+}
+
+// native_read_file : Read callback to decide whether to natively read or punt call to Go code
+static int native_read_file(char *path, char *buf, size_t size, off_t offset, fuse_file_info_t *fi)
+{
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    
+    if (handle_obj->fd == 0) {
+        return libfuse_read(path, buf, size, offset, fi);
+    }
+
+    return native_pread(path, buf, size, offset, handle_obj);
+}
+
+// native_write_file : Write callback to decide whether to natively write or punt call to Go code
+static int native_write_file(char *path, char *buf, size_t size, off_t offset, fuse_file_info_t *fi)
+{
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    
+    if (handle_obj->fd == 0) {
+        return libfuse_write(path, buf, size, offset, fi);
+    }
+    
+    return native_pwrite(path, buf, size, offset, handle_obj);
+}
+
+// native_flush_file : Flush the file natively and call flush up in the pipeline to upload this file
+static int native_flush_file(char *path, fuse_file_info_t *fi)
+{
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    int ret = libfuse_flush(path, fi);
+    if (ret == 0) {
+        // As file is flushed and uploaded, reset the dirty bit here
+        handle_obj->dirty = 0;
+    }
+
+    return ret;
+}
+
+
+#ifdef ENABLE_READ_AHEAD
```

Comment:
can we explicitly mark this code as still in progress?

Reply:
This code is disabled as of now so it will never hit. Just wanted to preserve some code done as part of read-ahead so that in future if we want to add it back, we have some reference. But the code is disabled so just treat this like some comments only.